ribbon-core/src/main/java/com/netflix/client/config/AbstractReloadableClientConfig.java

```diff
@@ -0,0 +1,389 @@
+ */
+public abstract class AbstractReloadableClientConfig implements IClientConfig {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractReloadableClientConfig.class);
+
+    private static final String DEFAULT_CLIENT_NAME = "";
+    private static final String DEFAULT_NAMESPACE = "ribbon";
+
+    // Map of raw property names (without namespace or client) to values set via code
+    private final Map<String, Object> defaultProperties = new HashMap<>();
+
+    // Map of all seen dynamic properties.  This map will hold on properties requested with the exception of
+    // those returned from getGlobalProperty().
+    private final Map<IClientConfigKey, ReloadableProperty<?>> dynamicProperties = new ConcurrentHashMap<>();
+
+    // List of actions to perform when configuration changes.  This includes both updating the Property instances
+    // as well as external consumers.
+    private final List<Runnable> changeActions = new CopyOnWriteArrayList<>();
+
+    private final AtomicLong refreshCounter = new AtomicLong();
+
+    private String clientName;
+
+    private String namespace = DEFAULT_NAMESPACE;
+
+    private boolean isDynamic;
+
+    public AbstractReloadableClientConfig() {
+        this(DEFAULT_CLIENT_NAME);
+        this.isDynamic = false;
+    }
+
+    public AbstractReloadableClientConfig(String clientName) {
+        this.clientName = clientName;
+        this.isDynamic = true;
+    }
+
+    /**
+     * Refresh all seen properties from the underlying property storage
+     */
+    public final void reload() {
+        changeActions.forEach(Runnable::run);
+        cachedToString = null;
+    }
+
+    public void setClientName(String clientName){
+        this.clientName  = clientName;
+    }
+
+    @Override
+    public final String getClientName() {
+        return clientName;
+    }
+
+    @Override
+    public String getNameSpace() {
+        return namespace;
+    }
+
+    @Override
+    public final void setNameSpace(String nameSpace) {
+        this.namespace = nameSpace;
+    }
+
+    @Override
+    public void loadProperties(String clientName) {
+        this.isDynamic = true;
+        this.clientName = clientName;
+    }
+
+    @Override
+    public void loadDefaultValues() {
+        isDynamic = true;
+    }
+
+    @Override
+    public final Map<String, Object> getProperties() {
+        Map<String, Object> result = new HashMap<>(dynamicProperties.size());
+
+        dynamicProperties.forEach((key, prop) ->
+            prop.getOptional().ifPresent(value -> result.put(key.key(), value.toString()))
+        );
+
+        LOG.info(result.toString());
+        return result;
+    }
+
+    /**
+     * Get a typed property value from the underlying storage mechanism.
+     * @param key
+     * @param type
+     * @param <T>
+     * @return Property value or null if not found
+     */
+    protected abstract <T> Optional<T> getFromStorage(String key, Class<T> type);
```

Comment:
maybe just "load"? The connotation here for me was that there would be some unusual property loading going on that skipped some part of resolution whereas something like "load" more implies it's the standard path to get something.